flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,594 @@
+                    meta = ReadWriteUtils.OBJECT_MAPPER.fromJson((String) row.getField(row.getArity() - 2), HashMap.class);
+                }
+            }
+            for (Object obj : broadcastVar) {
+                Row row = (Row) obj;
+                if (row.getField(FASTDISTANCE_TYPE_INDEX) != null) {
+                    long type = (long) row.getField(FASTDISTANCE_TYPE_INDEX);
+                    if (type == 1L) {
+                        dictData.add(
+                                FastDistanceMatrixData.fromString(
+                                        (String) row.getField(DATA_INDEX)));
+                    }
+                }
+            }
+            if (meta.containsKey(KnnParams.FEATURE_COLS.name)) {
+                selectedCols =
+                        ReadWriteUtils.OBJECT_MAPPER.fromJson(
+                                (String) meta.get(KnnParams.FEATURE_COLS.name), String[].class);
+            } else {
+                vectorCol =
+                        ReadWriteUtils.OBJECT_MAPPER.fromJson((String) meta.get(KnnParams.VECTOR_COL.name), String.class);
+            }
+
+            modelData = new KnnModelData(dictData, new EuclideanDistance());
+            idType =
+                    LogicalTypeDataTypeConverter.toDataType(
+                            LogicalTypeParser.parse((String) this.meta.get("idType")));
+        }
+    }
+
+    /**
+     * this operator use mapper to load the model data and do the prediction. if you want to write a
+     * prediction operator, you need implement a special mapper for this operator.
+     */
+    private static class PredictOperator
+            extends AbstractUdfStreamOperator<Row, RichMapFunction<Row, Row>>
+            implements OneInputStreamOperator<Row, Row> {
+
+        public PredictOperator(RichMapFunction<Row, Row> userFunction) {
+            super(userFunction);
+        }
+
+        @Override
+        public void processElement(StreamRecord<Row> streamRecord) throws Exception {
+            Row value = streamRecord.getValue();
+            output.collect(new StreamRecord<>(userFunction.map(value)));
+        }
+    }
+
+    private ResolvedSchema getOutputSchema(
+            ResolvedSchema dataSchema, Map<Param<?>, Object> params, DataType idType) {
+        String[] reservedCols = dataSchema.getColumnNames().toArray(new String[0]);
+        DataType[] reservedTypes = dataSchema.getColumnDataTypes().toArray(new DataType[0]);
+        String[] resultCols = new String[] {(String) params.get(KnnParams.PREDICTION_COL)};
+        DataType[] resultTypes = new DataType[] {idType};
+        return ResolvedSchema.physical(
+                ArrayUtils.addAll(reservedCols, resultCols),
+                ArrayUtils.addAll(reservedTypes, resultTypes));
+    }
+
+    /**
+     * cast data x to t type.
+     *
+     * @param x data.
+     * @param t type.
+     * @return
+     */
+    private static Object castTo(Object x, DataType t) {
```

Comment:
Do we really need this class or can this class be simplified? It looks a little bit dummy to do so many `if else`.

Reply:
OK, I will refine it later.

Reply:
done